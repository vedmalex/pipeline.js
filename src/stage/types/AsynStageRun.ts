import { z } from 'zod'
import { StageObject, StageObjectSchema } from './StageObject'

export type AsyncStageRun<R extends StageObject> = (err: unknown, context: unknown) => Promise<R>

export const AsyncStageRunSchema = z.function(z.tuple([z.unknown().optional(), StageObjectSchema]), z.promise(StageObjectSchema))
