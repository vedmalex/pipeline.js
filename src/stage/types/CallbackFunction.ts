import * as z from 'zod'
import { is_async_function } from './is_async_function'

export type CallbackFunction<R> = (err?: any, res?: R) => void

export const CallbackFunction = z.function().args(z.any().optional(), z.any().optional()).returns(z.void())

export function CallbackFunctionWrap(inp: unknown) {
  if (isCallbackFunction(inp)) {
    return CallbackFunction.implement(inp)
  } else {
    throw new Error('input not suitable for callback')
  }
}

export function isCallbackFunction<R>(inp?: unknown): inp is CallbackFunction<R> {
  return typeof inp === 'function' && !is_async_function(inp) && inp.length <= 2
}
