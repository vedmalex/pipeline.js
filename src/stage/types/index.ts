export * from './CallbackFunction'
export * from './StageEvaluateFunction'
export * from './StageRun'
export * from './ValidateFunction'
