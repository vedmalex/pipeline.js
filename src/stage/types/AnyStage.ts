import { z } from 'zod'
import { ContextProxySchema, ContextType } from '../Context'
import { StageConfigSchema } from '../StageConfig'
import { CallbackFunction, CallbackFunctionSchema } from './CallbackFunction'
import { StageObject, StageObjectSchema } from './StageObject'

export interface AnyStage<R extends StageObject> {
  config: unknown
  get reportName(): string
  get name(): string
  toString(): string
  execute<T extends StageObject>(context: unknown): Promise<T>
  execute<T extends StageObject>(context: unknown, callback: CallbackFunction<ContextType<R & T>>): void
  execute<T extends StageObject>(err: unknown, context: R, callback: CallbackFunction<ContextType<R & T>>): void
  execute<T extends StageObject>(
    _err?: unknown,
    _context?: R,
    _callback?: CallbackFunction<ContextType<R & T>>,
  ): void | Promise<T>
}

export const AnyStageSchema = z.object({
  config: StageConfigSchema,
  get reportName() {
    return z.string()
  },
  get name() {
    return z.string()
  },
  toString: z.function(z.tuple([]), z.string()),
  execute: z.union([
    z.function(
      z.tuple([
        ContextProxySchema,
      ]),
      z.promise(StageObjectSchema),
    ),
    z.function(
      z.tuple([
        ContextProxySchema,
        CallbackFunctionSchema,
      ]),
      z.void(),
    ),
    z.function(
      z.tuple([
        z.unknown(),
        ContextProxySchema,
        CallbackFunctionSchema,
      ]),
      z.void(),
    ),
  ]),
}).passthrough()
