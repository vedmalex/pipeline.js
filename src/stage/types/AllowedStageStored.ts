import { AnyStage } from './AnyStage'
import { RunPipelineFunction } from './RunPipelineFunction'

/**
 * хранимая часть конфига
 */
export type AllowedStageStored<Input, Output, Config> =
  | Config
  | RunPipelineFunction<Input, Output>
  | AnyStage<Input, Output>
