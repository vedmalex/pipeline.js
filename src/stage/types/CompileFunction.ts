import { AnyStage } from './AnyStage'
import { is_async_function } from './is_async_function'
import { StageRun } from './StageRun'

export type CompileFunction<R> = (this: AnyStage<R>, rebuild?: boolean) => StageRun<R>

export function isCompileFunction<R>(inp?: unknown): inp is CompileFunction<R> {
  return !is_async_function(inp) && typeof inp == 'function' && inp.length <= 1
}
