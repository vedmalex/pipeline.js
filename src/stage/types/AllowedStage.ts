import { StageConfig } from '../StageConfig'
import { AllowedStageStored } from './AllowedStageStored'

export type AllowedStage<R, CONFIG extends StageConfig<R>> = string | AllowedStageStored<R, CONFIG>
