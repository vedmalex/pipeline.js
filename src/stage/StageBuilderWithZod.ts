import { z } from 'zod'
import { ComplexError, CreateError } from './errors'
import { Stage } from './stage'
import { StageConfig } from './StageConfig'
import { CallbackFunction, isRunPipelineFunction, StageObject, StageRun, UnsetMarker } from './types'

/**
 * @internal
 * @see https://github.com/ianstormtaylor/superstruct/blob/7973400cd04d8ad92bbdc2b6f35acbfb3c934079/src/utils.ts#L323-L325
 */
export type Simplify<TType> = TType extends any[] | Date ? TType
  : { [K in keyof TType]: TType[K] }

export type OverwriteIfDefined<TType, TWith> = UnsetMarker extends TType ? TWith
  : Simplify<TType & TWith>

export type Merge<S, D> = Simplify<
  & { [K in keyof S]: K extends keyof D ? UnsetMarker extends D[K] ? S[K] : D[K] : S[K] }
  & { [K in keyof D]: K extends keyof S ? UnsetMarker extends S[K] ? D[K] : S[K] : D[K] }
>

export type Person = { name: string }
export type Employee = OverwriteIfDefined<{ name: UnsetMarker }, Person>
export type EmployeeSimpl = Merge<Person, Merge<{ empno: number }, { empno: UnsetMarker }>>

export interface BuilderParams {
  _stage: unknown
  _input: unknown
  _output: unknown
  _usage: {}
}

// упрощает работу с chain
export type InferParams<TParams extends BuilderParams, Usage extends keyof StageBuilder<TParams>> = {
  _stage: TParams['_stage']
  _input: TParams['_input']
  _output: TParams['_output']
  _usage: TParams['_usage'] & Pick<StageBuilder<TParams>, Usage>
}

export type AnyStageConfig = StageConfig<any>

export type BuilderDef<TStage> = {
  stage: TStage
  inputs: Parser
  outputs: Parser
  cfg: InferConfig<TStage>
}

export type InferConfig<TStage> = TStage extends Stage<any, infer $TConfig> ? $TConfig
  : TStage extends Stage<infer $Input, any> ? StageConfig<$Input>
  : {}

export type InferContext<TStage> = TStage extends Stage<infer $Input, any> ? $Input : UnsetMarker

export function stage<TStage extends Stage<any>>(
  _def: Partial<BuilderDef<TStage>> = {},
): StageBuilder<{
  _stage: TStage
  _input: UnsetMarker
  _output: UnsetMarker
  _usage: {}
}> {
  return {
    _def: _def as BuilderDef<TStage>,
    input(input) {
      if (!_def.cfg) {
        _def.cfg = {} as InferConfig<TStage>
      }
      _def.cfg.input = input
      return stage({
        ..._def,
        inputs: input,
      }) as any
    },
    output(output) {
      if (!_def.cfg) {
        _def.cfg = {} as InferConfig<TStage>
      }
      _def.cfg.output = output
      return stage({
        ..._def,
        outputs: output,
      }) as any
    },
    run(run) {
      if (!_def.cfg) {
        _def.cfg = {} as InferConfig<TStage>
      }
      if (run && isRunPipelineFunction(run)) {
        _def.cfg.run = run as any
      } else {
        throw CreateError('run should be a `RunPipelineFunction`')
      }
      return stage({
        ..._def,
      }) as any
    },
    name(name) {
      if (!_def.cfg) {
        _def.cfg = {} as InferConfig<TStage>
      }
      _def.cfg.name = name
      return stage({
        ..._def,
      }) as any
    },
    rescue(rescue) {
      if (!_def.cfg) {
        _def.cfg = {} as InferConfig<TStage>
      }
      _def.cfg.rescue = rescue as any
      return stage({
        ..._def,
      }) as any
    },
    compile(compile) {
      if (!_def.cfg) {
        _def.cfg = {} as InferConfig<TStage>
      }
      _def.cfg.compile = compile as any
      return stage({
        ..._def,
      }) as any
    },
    build() {
      return new Stage<InferContext<TStage>, InferConfig<TStage>>(
        (_def as BuilderDef<TStage>).cfg as InferConfig<TStage>,
      ) as any
    },
  }
}

export type ParserZod<TInput, TParsedInput> = {
  _input: TInput
  _output: TParsedInput
}

// export type Parser = ParserZod<any, any>
export type Parser = z.ZodTypeAny

export type inferParser<TParser extends Parser> = TParser extends ParserZod<infer $TIn, infer $TOut> ? {
    in: $TIn
    out: $TOut
  }
  : UnsetMarker

type ErrorMessage<TMessage extends string> = TMessage

type InferKeys<T> = T extends object ? keyof T : never

// берет тип на основании трёх параметров
//
type SchemaType<
  TParams extends BuilderParams,
  $Parser extends Parser,
  key extends keyof TParams,
  parser extends keyof inferParser<$Parser>,
> = TParams[key] extends UnsetMarker ? $Parser
  : inferParser<$Parser>[parser] extends Record<string, unknown> | undefined
    ? TParams[key] extends Record<string, unknown> | undefined ? undefined extends inferParser<$Parser>[parser] // if current is optional the previous must be too
        ? undefined extends TParams[key] ? $Parser
        : ErrorMessage<'Cannot chain an optional parser to a required parser'>
      : $Parser
    : ErrorMessage<'All input parsers did not resolve to an object'>
  : ErrorMessage<'All input parsers did not resolve to an object'>

export interface StageBuilder<TParams extends BuilderParams> {
  /**
   * input for stage
   * @param schema only object allowed
   */
  input<$Parser extends Parser>(
    schema: SchemaType<TParams, $Parser, '_input', 'in'>,
  ): Omit<
    StageBuilder<
      Merge<
        InferParams<TParams, 'input'>,
        {
          _input: OverwriteIfDefined<
            TParams['_input'],
            inferParser<$Parser>['in']
          >
        }
      >
    >,
    InferKeys<TParams['_usage']> | 'input'
  >
  /**
   * output of stage
   * @param schema only object allowed
   */
  output<$Parser extends Parser>(
    schema: SchemaType<TParams, $Parser, '_output', 'out'>,
  ): Omit<
    StageBuilder<
      Merge<
        InferParams<TParams, 'output'>,
        {
          _output: OverwriteIfDefined<
            TParams['_output'],
            inferParser<$Parser>['in']
          >
        }
      >
    >,
    InferKeys<TParams['_usage']> | 'ouput'
  >
  run<$Input = TParams['_input']>(fn: CustomRun<$Input>): Omit<
    StageBuilder<
      InferParams<TParams, 'run'>
    >,
    InferKeys<TParams['_usage']> | 'run'
  >
  name(name: string): Omit<
    StageBuilder<
      InferParams<TParams, 'name'>
    >,
    InferKeys<TParams['_usage']> | 'name'
  >
  rescue(rescue: Rescue<TParams['_input']>): Omit<
    StageBuilder<
      InferParams<TParams, 'rescue'>
    >,
    InferKeys<TParams['_usage']> | 'rescue'
  >
  compile(fn: Compile<TParams['_input']>): Omit<
    StageBuilder<
      InferParams<TParams, 'compile'>
    >,
    InferKeys<TParams['_usage']> | 'compile'
  >
  build(): Stage<
    TParams['_output'] extends UnsetMarker ? TParams['_input'] : TParams['_output'],
    StageConfig<TParams['_output'] extends UnsetMarker ? TParams['_input'] : TParams['_output']>
  >
  _def: BuilderDef<TParams['_stage']>
}

export type CustomRun<$P1> = (
  this: $P1 extends StageObject ? $P1 : never,
  p1?: ComplexError | $P1 | undefined,
  p2?: $P1 | undefined,
  p3?: CallbackFunction<$P1> | undefined,
) => Promise<$P1> | $P1 | void

export type Rescue<$P1> = (
  this: $P1 extends StageObject ? $P1 : never,
  p1?: Error | $P1 | undefined,
  p2?: $P1 | undefined,
  p3?: CallbackFunction<$P1> | undefined,
) => Promise<$P1> | $P1 | void

export type Ensure<$P1> = (
  p1?: $P1 extends StageObject ? $P1 : never,
  p2?: CallbackFunction<$P1> | undefined,
) => Promise<$P1> | $P1 | void

export type ValidateFn<$P1> = (
  this: $P1 extends StageObject ? $P1 : never,
  p1?: $P1 | undefined,
  p2?: CallbackFunction<boolean> | undefined,
) => Promise<boolean> | boolean | void

export type Compile<TStage> = (
  this: TStage extends Stage<infer $P, any> ? StageRun<$P> : never,
  rebuild?: boolean,
) => TStage extends Stage<infer $P, any> ? StageRun<$P> : never

export type Precompile<$P1, TStage> = (
  this: TStage extends Stage<$P1> ? TStage : never,
) => void

// const st = createBuilder()
//   .input(z.object({ name: z.string() }))
//   .output(z.object({ full: z.string() }))
//   .name('stage')
//   .rescue(function (err, ctx) {
//   })
//   .ensure((ctx, done) => {
//     return ctx ?? ctx
//   })
//   .run(function () {
//     this
//   })
//   .build()

// const st2 = createBuilder()
// .input(z.object({ name: z.string() }))
// .output(z.object({ full: z.string() }))
// .name('stage')
// .rescue(function (err, ctx) {
// })
// .ensure((ctx, done) => {
//   return ctx ?? ctx
// })
// .run(function () {
//   this
// })

// st2.build()
