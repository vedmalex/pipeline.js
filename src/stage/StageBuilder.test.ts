import 'jest'
import { z } from 'zod'
import { stage } from './StageBuilderWithZod'
import { makeCallbackArgs } from './types'

describe('stageBuilder', () => {
  // дальше работаем с типами!!! чтобы был контроль входщих данных и выходящих
  // исправить работу с Stage<any> похоже что TStage не нужен
  it('run', done => {
    const st = stage()
      .input(z.object({ name: z.string().optional() }))
      .output(z.object({ name: z.string() }))
      .run(ctx => {
        if (ctx) {
          ctx.name = 'name'
        }
      })
      .build()

    st.execute({}, (err, res) => {
      expect(err).toBeUndefined()
      expect(res).toMatchObject({ name: 'name' })
      done()
    })
  })
  it('create named', () => {
    const s = stage()
      .name('names stage').build()
    expect(s).not.toBeNull()
    expect(s).toMatchSnapshot('names stage')
  })
  it('create with function', () => {
    const s = stage()
      .input(z.object({ name: z.string().optional() }))
      .run(function (this) {
        this.name = 'run this Stage'
      }).build()

    expect(s).not.toBeNull()
    expect(s).toMatchSnapshot('function stage')
  })

  it('create with Lambda 3', () => {
    const s = stage()
      .input(z.object({ name: z.string().optional() }))
      .output(z.object({ name: z.string() }))
      .run((err, ctx, done) => {
        if (!err && ctx && done) {
          ctx.name = 'run the stage'
          done(makeCallbackArgs(undefined, ctx))
        } else if (done) {
          done(makeCallbackArgs(err))
        }
      }).build()
    expect(s).toMatchSnapshot('lambda stage')
  })

  it('intialize using schema and validate separately', () => {
    expect(() =>
      stage()
        .input(z.object({ name: z.string() }))
        .run(() => {})
        .build()
    ).not.toThrow()
    expect(() =>
      stage()
        .run(() => {})
        .input(z.object({}).passthrough())
        .build()
    ).not.toThrow()
  })

  it('validate using schema', done => {
    const st = stage()
      .run(() => {})
      .input(z.object({}).passthrough())
      .output(z.object({ name: z.string() }))
      .build()

    st.execute({ fullname: 1 }, (err, res) => {
      expect(err).not.toBeUndefined()
      expect(err).toMatchSnapshot()
      done()
    })
  })

  it('initialize other stuff sucessfully', () => {
    let st = stage()
      .name('stage')
      .run(() => {})
      .rescue(() => {})
      .build()

    expect(st.name).toBe('stage')
    expect(st.reportName).toBe(`STG:${st.name}`)
    expect(st).toMatchSnapshot('schema stage 1')
  })
  it('create with Lambda 2', () => {
    const s = stage()
      .input(z.object({ name: z.string() }))
      .run((err, ctx) => {
        if (!err && ctx) {
          ctx.name = 'run the stage'
          return Promise.resolve(ctx)
        } else {
          return Promise.reject(err)
        }
      })
    expect(s).toMatchSnapshot('lambda stage')
  })

  it('create with Config', () => {
    const s = stage()
      .input(z.object({name: z.string()}))
      .run((err, ctx, done) => {
        if (done) {
          if (!err && ctx) {
            ctx.name = 'run the stage'
            done(makeCallbackArgs(undefined, ctx))
          } else {
            done(makeCallbackArgs(err))
          }
        }
      })
    expect(s).toMatchSnapshot('config stage')
  })

  // it('use Precompile to run everything', () => {
  //   type CTX = {
  //     name: string
  //   }
  //   interface Config extends StageConfig<CTX, CTX> {
  //     stages: Array<IStage<any, any> | RunPipelineFunction>
  //     addStage(
  //       this: Config,
  //       stage:
  //         | StageConfig<CTX, CTX>
  //         | SingleStageFunction
  //         | IStage<any, any>,
  //     ): void
  //   }

  //   const pipeline = new Stage<CTX, CTX, Config>({
  //     stages: [],
  //     precompile(this: Config) {
  //       let run: StageRun<CTX, CTX> = (
  //         err: Possible<ComplexError>,
  //         context: CTX,
  //         done: CallbackFunction,
  //       ) => {
  //         let i = -1
  //         //sequential run;
  //         let next = (err: Possible<ComplexError>, context: CTX | undefiend) => {
  //           i += 1
  //           if (i < this.stages.length) {
  //             run_or_execute(this.stages[i], err, context, next)
  //           } else if (i == this.stages.length) {
  //             done(err, context)
  //           } else {
  //             done(new Error('done call more than once'), context)
  //           }
  //         }
  //         next(err, context)
  //       }

  //       if (this.stages.length > 0) {
  //         this.run = run
  //       } else {
  //         this.run = empty_run
  //       }
  //     },
  //     addStage(
  //       _stage:
  //         | StageConfig<CTX, CTX>
  //         | RunPipelineFunction
  //         | IStage<any, any>,
  //     ) {
  //       let stage: IStage<any, any> | RunPipelineFunction| undefined
  //       if (typeof _stage === 'function') {
  //         stage = _stage
  //       } else {
  //         if (typeof _stage === 'object') {
  //           if (!isIStage(_stage)) {
  //             stage = new Stage(_stage)
  //           } else {
  //             stage = _stage
  //           }
  //         }
  //       }
  //       if (stage) {
  //         this.stages.push(stage)
  //         this.run = undefined
  //       }
  //     },
  //   })
  // })

  // it('rescue do not enter twice into single rescue block', () => {
  //   let rescue = 0
  //   var st = new Stage({
  //     run: function (err, ctx, done) {
  //       throw new Error('error')
  //     },
  //     rescue: function (err, conext) {
  //       rescue += 1
  //       if (err.payload[0].message !== 'some') {
  //         throw err
  //       }
  //     },
  //   })
  //   st.execute({}, (err, res) => {
  //     expect(rescue).toBe(1)
  //   })
  // })
})
