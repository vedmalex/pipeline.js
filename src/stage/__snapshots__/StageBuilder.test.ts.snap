// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stageBuilder create named: names stage 1`] = `
Stage {
  "_config": {
    "name": "names stage",
  },
  "run": undefined,
  [Symbol(stage)]: true,
}
`;

exports[`stageBuilder create with Config: config stage 1`] = `
Stage {
  "_config": {
    "name": 
"(err, ctx, done) => {
      if (!err)
        ctx.name = "run the stage", done(void 0, ctx);
      else
        done(err);
    }"
,
    "run": [Function],
  },
  "run": undefined,
  [Symbol(stage)]: true,
}
`;

exports[`stageBuilder create with Lambda 2: lambda stage 1`] = `
Stage {
  "_config": {
    "name": 
"(err, ctx) => {
      if (!err)
        return ctx.name = "run the stage", Promise.resolve(ctx);
      else
        return Promise.reject(err);
    }"
,
    "run": [Function],
  },
  "run": undefined,
  [Symbol(stage)]: true,
}
`;

exports[`stageBuilder create with Lambda 3: lambda stage 1`] = `
Stage {
  "_config": {
    "name": 
"(err, ctx, done) => {
      if (!err)
        ctx.name = "run the stage", done(void 0, ctx);
      else
        done(err);
    }"
,
    "run": [Function],
  },
  "run": undefined,
  [Symbol(stage)]: true,
}
`;

exports[`stageBuilder create with function: function stage 1`] = `
Stage {
  "_config": {
    "name": 
"function() {
      this.name = "run this Stage";
    }"
,
    "run": [Function],
  },
  "run": undefined,
  [Symbol(stage)]: true,
}
`;

exports[`stageBuilder initialize other stuff sucessfully: schema stage 1 1`] = `
Stage {
  "_config": {
    "name": "stage",
    "run": [Function],
  },
  "run": undefined,
  [Symbol(stage)]: true,
}
`;

exports[`stageBuilder validate using schema 1`] = `[ComplexError]`;
